src/flow/src/expr/scalar.rs

```diff
@@ -105,25 +105,52 @@ impl ScalarExpr {
     /// This method is applicable even when `permutation` is not a
     /// strict permutation, and it only needs to have entries for
     /// each column referenced in `self`.
-    pub fn permute(&mut self, permutation: &[usize]) {
+    pub fn permute(&mut self, permutation: &[usize]) -> Result<(), EvalError> {
+        if self
+            .get_all_ref_columns()
+            .into_iter()
+            .any(|i| i >= permutation.len())
+        {
+            return InvalidArgumentSnafu {
+                reason: format!(
+                    "permutation {:?} is not a valid permutation for expression {:?}",
+                    permutation, self
+                ),
+            }
+            .fail();
+        }
```

Comment:
how about letting `visit_*` return `Result<(), EvalError>`? Then we can save some memory allocation and traversal.

Reply:
> how about letting `visit_*` return `Result<(), EvalError>`? Then we can save some memory allocation and traversal.

Just realized that:
1. It will cause we end up with a partially permuted expression, which I don't think can be properly recovered
2. permute& other similiar function is a cold path, like it will only run once if a query need to be compiled/built, and any following new data arrive would never call those functions, so a little overhead is acceptable